New timeline API

jplatte opened the pull request:
WIP: For now, this is mostly a data model / stiching together of event handlers to feed the model, actually making this API consumable for users is still entirely a TODO, an the same goes for documentation.

jplatte commented:
Things I would like to discuss at the current stage:

* What do we want to do with the existing `experimental-timeline` flag?
  * We have an example plus the FFI crate activating it, for IDE experience to not be horrible when working on the new one, those need to allow both features to be active at the same time or make those dependencies optional (or possibly just the FFI one)
* How do we deal with thing like replies
 * I'm pretty sure we want users to be able to render the reply, even if they can't render the replied-to message yet
 * One approach: Don't make the replied-to message part of the timeline item, but offer a separate async function to obtain the model bits required to render it
 * Another approach: Make the replied-to message part of the timeline item, but have three states: [no reply, reply loading, reply] and automatically obtain the required model bits when encountering a reply, then update the timeline item again
 * The same question applies to media, although the answer could be different (the first approach seems more reasonable there because you might want to do the media loading at different times depending on network conditions)

poljar commented:
> Things I would like to discuss at the current stage:
> 
>     * What do we want to do with the existing `experimental-timeline` flag?
>       
>       * We have an example plus the FFI crate activating it, for IDE experience to not be horrible when working on the new one, those need to allow both features to be active at the same time or make those dependencies optional (or possibly just the FFI one)

I think we should just remove the feature from the codebase. The example should be updated to use the new feature. For the FFI I would expect the FFI crate to be updated and hopefully tested in Element X as part of this PR. This does mean some coordination with the Element X team and live testing but should overall assure us that the PR works and is what Element X wants.

>       * One approach: Don't make the replied-to message part of the timeline item, but offer a separate async function to obtain the model bits required to render it
>       * Another approach: Make the replied-to message part of the timeline item, but have three states: [no reply, reply loading, reply] and automatically obtain the required model bits when encountering a reply, then update the timeline item again
>       * The same question applies to media, although the answer could be different (the first approach seems more reasonable there because you might want to do the media loading at different times depending on network conditions)

In both cases you can't be too sure if the client actually wants to actually render the individual item, i.e. is the message in view or not. Seems like any decision on whether we should fetch additional data from the server or not should be delegated to the client itself.

In both cases the client can and should display a placeholder until it has all the data to render things fully.

jplatte commented:
Re. `experimental-timeline`, I was kind of hoping for this PR to not get huge, i.e. merge it before we add any FFI things. But there's definitely merit to getting it close to feature parity before merging, and replacing the old API immediately.

> In both cases you can't be too sure if the client actually wants to actually render the individual item, i.e. is the message in view or not. Seems like any decision on whether we should fetch additional data from the server or not should be delegated to the client itself.

This is a good point. I guess we can still have the three-states thing though, have the client call `timeline.load_reply(event_id)` which doesn't return anything, but causes another update to that timeline item to be sent once the data is available. The same thing would work for media, potentially with separate calls for getting the thumbnail, and the full thing (if the latter even makes sense to show inline, but I guess sometimes it does).

zecakeh commented:
We need the `TimelineKey`, `TimelineItemContent`, `ReactionDetails`, `Message` and `PaginationOutcome` types to be exported by the `matrix-sdk` crate for this to be usable :slightly_smiling_face: .

jplatte commented:
Right. I've now removed the re-exports from the `room` module and made the `room::timeline` module public instead, since I don't think we should be cluttering the `room` module with all those symbols.

zecakeh commented:
I'm guessing all timeline event types will be supported later? What is the plan with events that are not successfully deserialized?

jplatte commented:
Yeah, the plan is to support all specced event types. Events that aren't successfully deserialized haven't been discussed so far, but I'm pretty sure we'll want to cover that somehow. Element Web has a dev mode thing where reactions (and I think edits too) show up as greyed-out extra entries in the timeline, I think we'll want to support something like that, maybe in the form of a "raw" timeline item that is added whenever an event doesn't otherwise produce a new timeline item.

jplatte commented:
I want to see whether bindings are generating correctly, but this isn't actually ready for review.

zecakeh commented:
This is currently implemented in this MR on Fractal: [fractal!1136](https://gitlab.gnome.org/GNOME/fractal/-/merge_requests/1136).

I'll try to update it frequently with new changes from this PR.

jplatte commented:
Rebased on top of #1000 and cleaned up the implementation a bit (made possible by that PR and others). Also added support for accessing the raw JSON of the initial event. How to deal with the JSON of aggregated events is left as a FIXME, potentially one that won't be solved in this PR.

pixlwave commented on bindings/matrix-sdk-ffi/src/api.udl:
```diff
@@ -112,57 +106,179 @@ interface Room {
+
+interface EventTimelineItem {
+    TimelineKey key();
     string sender();
```
Would it be possible to have this `sender` provide the displayname and avatar URL so we don't need to fetch those separately? E.g. something along the lines of
```
struct SenderDetails {
    let id: String
    let displayname: String
    let avatarURL: URL
}
```

jplatte replied:
This was discussed a while back and decided against, partially to keep the API simpler and partially because clients would usually want to have one mapping of user ID to displayname / avatar per room that's used both for the timeline and member list.

pixlwave replied:
> one mapping of user ID to displayname / avatar per room

A counter-point to this. Currently in Element, we show the name/avatar of the user at the point in time when they sent the event. So having a single mapping wouldn't allow that to be the case.

jplatte replied:
Yeah, that's a known trade-off, it's noted in https://hdoc.ergaster.org/IFvYg5aSSkaBizYFmVlgIA. Unfortunately I don't have a permalink to the previous discussion about this, but maybe we can just re-discuss.

pixlwave commented on bindings/matrix-sdk-ffi/src/api.udl:
```diff
@@ -112,57 +106,179 @@ interface Room {
+    TimelineKey key();
     string sender();
+    TimelineItemContent content();
+    sequence<Reaction> reactions();
```
Ideally, these would be aggregated reactions on the timeline API. (At a later date a separate method to request individual reactions for a specific event would be useful to show more details but we don't particularly need that in the timeline itself).

That said, looking at the checklist on the PR I think this is known :)

jplatte replied:
Yeah, this doesn't work for pagination because there you don't have the full details. I think it should still be in the data model such that hovers over the symbols can be implemented easily. It will just not be available until you fetch all details of the event.

pixlwave commented on bindings/matrix-sdk-ffi/src/api.udl:
```diff
@@ -112,57 +106,179 @@ interface Room {
+    string sender;
+};
+
+interface VirtualTimelineItem {};
```
I'm guessing this is intentionally not exposed yet, but just pointing out that it doesn't provide anything useable yet.

jplatte replied:
Yeah, this is just a placeholder. It doesn't make much of a difference whether I expose it until the timeline actually emits these items :)

pixlwave commented on crates/matrix-sdk/src/room/timeline/virtual_item.rs:
```diff
@@ -0,0 +1,8 @@
+#[derive(Clone, Debug)]
+pub enum VirtualTimelineItem {
+    /// A divider between messages of two days.
+    DayDivider,
```
It would be nice if this included the date of which day it represents (again maybe coming, but just thinking out loud here).

pixlwave commented on bindings/matrix-sdk-ffi/src/timeline.rs:
```diff
@@ -0,0 +1,409 @@
+impl TimelineItemContent {
+    pub fn as_message(self: Arc<Self>) -> Option<Arc<Message>> {
+        use matrix_sdk::room::timeline::TimelineItemContent as C;
+        unwrap_or_clone_arc_into_variant!(self, .0, C::Message(msg) => Arc::new(Message(msg)))
```
Double checking for my own knowledge, is this and others like `as_text`/`as_image` an expensive operation if it gets called repeatedly?

jplatte replied:
Unfortunately, I expect it to clone the entire inner message every time. Whether that fits your definition of expensive, I don't know ^^

pixlwave commented on bindings/matrix-sdk-ffi/src/timeline.rs:
```diff
@@ -0,0 +1,409 @@
+#[derive(Clone)]
+pub struct EventTimelineItem(matrix_sdk::room::timeline::EventTimelineItem);
+
+impl EventTimelineItem {
```
Last 2 thoughts for now - in EX, we're adding 2 more properties around these: `isOutgoing` (could maybe be named better but means was it sent by the current user) and `shouldShowSenderDetails` which is `false` if the previous item was sent by the same sender on the same day within a certain time threshold (again could probably be named better as its indicating a continual section of messages).

Not sure if these are too specific for the Rust side but mentioning as they would need to be re-implemented in EA too, so potentially worth considering?

jplatte replied:
Those both sound worth adding 👍🏼

pixlwave replied:
Thinking about this, the `isOutgoing` property might also be a good place to include the transaction ID so its possible to identify when an event that is currently onscreen as a local echo gets replaced by the echo from the server.

pixlwave reviewed (commented):
So overall this has been a breeze to integrate into ElementX, thanks 🙌

I've added some minor thoughts inline that I've had whilst doing so.

pixlwave commented on bindings/matrix-sdk-ffi/src/api.udl:
```diff
@@ -112,57 +106,179 @@ interface Room {
+    //
+    // Raises an exception if there are no timeline listeners.
+    [Throws=ClientError]
+    PaginationOutcome paginate_backwards();
```
The previous version of this method had a `u64 count` parameter. Would be be able to add that to this version too? Not necessarily now, just highlighting this as when the app is running on a Mac with a large display we would ideally want to paginate by a much larger number of events than when running on an old iPhone with a tiny screen.

jplatte replied:
Sure.

jplatte commented:
I want to see how much coverage the single test amounts to, even if it's not yet fully working. So this isn't actually ready for review, I just want CI to run..

jplatte commented:
This is now ready for review!

zecakeh commented:
Like I said on matrix, currently we don't get the aggregated reactions. However if we forward the `Relations` from the event to the `NewEventTimelineItem`, it works, since the rest of the code is in place.

poljar commented on crates/matrix-sdk/src/room/timeline/event_handler.rs:
```diff
@@ -0,0 +1,460 @@
+    End,
+}
+
+struct EventHandler<'a> {
```
Oh no, now we have two things called `EventHandler` in the codebase? Is there a reason for this to exist except to keep the number of methods on the `TimelineInner` struct slim?

poljar commented on crates/matrix-sdk/src/room/timeline/event_handler.rs:
```diff
@@ -0,0 +1,460 @@
+            _ => {}
+        }
+
+        if !self.event_added {
```
Can't this be handled like the `did_update` flag? Where we return the value instead of mutating things in some magic place somewhere else?

jplatte replied:
Well, there is only a single place where it is set, so I think it's still obvious enough what happens and this way the method signatures are shorter. We can try out the other way though.

poljar commented on crates/matrix-sdk/src/room/timeline/mod.rs:
```diff
@@ -0,0 +1,217 @@
+    }
+}
+
+fn find_event(
```
Hmm, don't we expect this to be problematic? Do we ever remove items from the timeline? I don't think we do. How many items do we want to support? Don't we want to let the caller configure how many items we keep?

We'll likely need an index that maps from the key to the index in any case. Let's at least add a fixme.

jplatte replied:
I'm a little confused, do you mean this is problematic if we keeps those indices around? The intention of this function is just to find the right index to update, it is only meant to be valid for a very short duration.

We could have a separate index so we don't have to do a linear scan, but that would be more complex and I don't think we need it right now (maybe never if we put a reasonable upper bound on the amount of entries that can be in the timeline).

poljar replied:
No, I meant the growth of the `MutableVec` is unbounded thus the `rfind()` and in turn the `find_event()` function will take an increasing amount of time in the worst case.

poljar commented on crates/matrix-sdk/src/room/timeline/mod.rs:
```diff
@@ -0,0 +1,217 @@
+        }
+    }
+
+    /// Add more events to the start of the timeline.
```
I wonder if calling this operation *scrolling* would add more clarity.

jplatte replied:
🤷🏼

poljar commented on crates/matrix-sdk/src/room/timeline/mod.rs:
```diff
@@ -0,0 +1,217 @@
+    ///
+    /// See [`SignalVecExt`](futures_signals::signal_vec::SignalVecExt) for a
+    /// high-level API on top of [`SignalVec`].
+    pub fn signal(&self) -> impl SignalVec<Item = Arc<TimelineItem>> {
```
Should we expose `SignalVec` here, if we just expose `impl Stream...` that would likely be more familiar to most people. Another thing, when I tried the API out myself, the stream was the nicer one to use I would say.

Do we lose much if we return the stream here instead?

jplatte replied:
I don't think we lose much, but also *why not both*? I'll add a `.stream()` accessor.

poljar commented on crates/matrix-sdk/src/room/timeline/event_handler.rs:
```diff
@@ -0,0 +1,460 @@
+
+        self.maybe_update_timeline_item(event_id, "edit", |item| {
+            if self.meta.sender != item.sender() {
+                info!(
```
All the log lines seem to be missing the room id, this will be quite annoying to debug if we don't know the room the log line applies to.

jplatte replied:
I've added it for local echo and back-pagination, for live events we should just add a useful span for all event handlers, not just the timeline ones.

poljar reviewed (approved):
Please add the license headers, the rest of my comments can wait for another PR, like the comments from my initial review.

gnunicorn commented on crates/matrix-sdk/src/room/timeline/event_handler.rs:
```diff
@@ -0,0 +1,460 @@
+        }
+
+        let mut lock = self.timeline.items.lock_mut();
+        if let Some((idx, item)) = find_event(&lock, event_id) {
```
I wonder if that could become a bottle neck if we have scrolled back a room for a bit ... but I'll put that off as a future problem.

gnunicorn reviewed (approved):
Aside from docs, and the big amount of FIXMEs and ToDos I'd prefer to see in our issue tracker rather than in some code comments, this looks good to me as the first iteration! Good job, well done!

jplatte commented on crates/matrix-sdk/src/room/timeline/mod.rs:
```diff
@@ -0,0 +1,217 @@
+    inner: TimelineInner,
+    room: room::Common,
+    start_token: Mutex<Option<String>>,
+    _end_token: Mutex<Option<String>>,
```
Yeah, future forwards pagination support.

jplatte commented on crates/matrix-sdk/src/room/common.rs:
```diff
@@ -251,6 +253,12 @@ impl Common {
         self.client.add_room_event_handler(self.room_id(), handler)
     }
 
+    #[allow(missing_docs)] // FIXME(jplatte)
```
Yeah, that was meant to be fixed before merge.

jplatte commented on crates/matrix-sdk/src/room/timeline/event_item.rs:
```diff
@@ -0,0 +1,305 @@
+
+    /// The reactions.
+    ///
+    /// Key: The reaction, usually an emoji.\
```
This is actually load-bearing, it inserts a line break before the `Value` line without putting them on entirely separate paragraphs (the same way as double space at the end of the line, but that is stripped by many editors).

jplatte commented on crates/matrix-sdk/src/room/timeline/event_handler.rs:
```diff
@@ -0,0 +1,460 @@
+    TimelineKey,
+};
+
+pub(super) fn handle_live_event(
```
`TimelineInner` is a fairly recent addition compared to the rest of this module but yeah it sounds reasonable.
